README.rst
======
esExporter
======

A CLI tool for exporting data from Elasticsearch into a file
----------------------------------------------------------------

Baseed on https://github.com/taraslayshchuk/es2csv developed. Command line utility, for querying Elasticsearch(7.x) in Lucene query syntax or Query DSL syntax and exporting result as documents into a file with custom fields.

Requirements
------------
| This tool should be used with Elasticsearch 7.x version.
| You also need `Python 2.7.x <https://www.python.org/downloads/>`_ and `pip <https://pip.pypa.io/en/stable/installing/>`_.

Installation
------------

From source:

.. code-block:: bash

    $ git clone https://github.com/AkitoZz/esExporter.git && python setup.py install


Usage
-----
.. code-block:: bash

 $ esExporter [-h] -q QUERY [-u URL] [-a AUTH] [-i INDEX [INDEX ...]] -o FILE
          [-f FIELDS [FIELDS ...]] [-S FIELDS [FIELDS ...]] [-d DELIMITER]
          [-m INTEGER] [-s INTEGER] [--verify-certs]
          [--ca-certs CA_CERTS] [--client-cert CLIENT_CERT]
          [--client-key CLIENT_KEY] [-v] [--debug] [--header-on]

 Arguments:
  -q, --query QUERY                        Query string in Lucene syntax.               [required]
  -o, --output-file FILE                   file location.                               [required]
  -u, --url URL                            Elasticsearch host URL. Default is http://localhost:9200.
  -a, --auth                               Elasticsearch basic authentication in the form of username:password.
  -i, --index-prefixes INDEX [INDEX ...]   Index name prefix(es). Default is ['logstash-*'].
  -f, --fields FIELDS [FIELDS ...]         List of selected fields in output. Default is ['_all'].
  -S, --sort FIELDS [FIELDS ...]           List of <field>:<direction> pairs to sort on. Default is [].
  -d, --delimiter DELIMITER                Delimiter to use in CSV file. Default is ",".
  -m, --max INTEGER                        Maximum number of results to return. Default is 0.
  -s, --scroll-size INTEGER                Scroll size for each batch of results. Default is 100.
  --verify-certs                           Verify SSL certificates. Default is False.
  --ca-certs CA_CERTS                      Location of CA bundle.
  --client-cert CLIENT_CERT                Location of Client Auth cert.
  --client-key CLIENT_KEY                  Location of Client Cert Key.
  -v, --version                            Show version and exit.
  --debug                                  Debug mode on.
  -h, --help                               show this help message and exit
  --header-on                              Export file with header

[ `Usage Examples <./docs/EXAMPLES.rst>`_ | `Release Changelog <./docs/HISTORY.rst>`_ ]


docs/HISTORY.rst
.. :changelog:

Release Changelog（developed based on https://github.com/taraslayshchuk/es2csv 5.5.2）
=================

7.5.1 (2022-03-03)
------------------
- Delete -D -t -k -r -e argument.
- Added --header-on argument for exported file with header or not.
- Updateing version elasticsearch to 7.x

5.5.2 (2018-03-21)
------------------
- Fixed encoding in field name to UTF-8. (Issue `#35 <https://github.com/taraslayshchuk/es2csv/issues/35>`_)
- Added --sort(-S) argument for sorting data by selected field. (Issue `#41 <https://github.com/taraslayshchuk/es2csv/issues/41>`_)
- Added requirement for version of python 2.7.*. (Issue `#8 <https://github.com/taraslayshchuk/es2csv/issues/8>`_, `#12 <https://github.com/taraslayshchuk/es2csv/issues/12>`_, `#20 <https://github.com/taraslayshchuk/es2csv/issues/20>`_, `#29 <https://github.com/taraslayshchuk/es2csv/issues/29>`_, `#33 <https://github.com/taraslayshchuk/es2csv/issues/33>`_ and `#38 <https://github.com/taraslayshchuk/es2csv/issues/38>`_)
- Update documentation with examples.
- Updating version elasticsearch-py to 5.5.*.


esExporter.py
import os
import gzip
import signal
import time
import json
import codecs
import elasticsearch
import progressbar
import thread
from functools import wraps


FLUSH_BUFFER = 1000  # Chunk of docs to flush in temp file
CONNECTION_TIMEOUT = 120
TIMES_TO_TRY = 3
RETRY_DELAY = 60


# Retry decorator for functions with exceptions
def retry(ExceptionToCheck, tries=TIMES_TO_TRY, delay=RETRY_DELAY):
    def deco_retry(f):
        @wraps(f)
        def f_retry(*args, **kwargs):
            mtries = tries
            while mtries > 0:
                try:
                    return f(*args, **kwargs)
                except ExceptionToCheck as e:
                    print(e)
                    print('Retrying in {} seconds ...'.format(delay))
                    time.sleep(delay)
                    mtries -= 1
                else:
                    print('Done.')
            try:
                return f(*args, **kwargs)
            except ExceptionToCheck as e:
                print('Fatal Error: {}'.format(e))
                exit(1)

        return f_retry

    return deco_retry


class EsExporter:

    def __init__(self, opts):
        self.opts = opts

        self.num_results = 0
        self.scroll_ids = []
        self.scroll_time = '2m'
        self.fields = []
        self.file = opts.output_file
        
        for sig in [signal.SIGINT, signal.SIGTERM]:
            signal.signal(sig, self.exit_by_signal)

    @retry(elasticsearch.exceptions.ConnectionError, tries=TIMES_TO_TRY)
    def create_connection(self):
        es = elasticsearch.Elasticsearch(self.opts.url, timeout=CONNECTION_TIMEOUT, http_auth=self.opts.auth,
                                         verify_certs=self.opts.verify_certs, ca_certs=self.opts.ca_certs,
                                         client_cert=self.opts.client_cert, client_key=self.opts.client_key)
        es.cluster.health()
        self.es_conn = es

    @retry(elasticsearch.exceptions.ConnectionError, tries=TIMES_TO_TRY)
    def check_indexes(self):
        indexes = self.opts.index_prefixes
        if '_all' in indexes:
            indexes = ['_all']
        else:
            indexes = [index for index in indexes if self.es_conn.indices.exists(index)]
            if not indexes:
                print('Any of index(es) {} does not exist in {}.'.format(', '.join(self.opts.index_prefixes), self.opts.url))
                exit(1)
        self.opts.index_prefixes = indexes

    @retry(elasticsearch.exceptions.ConnectionError, tries=TIMES_TO_TRY)
    def search_query(self):
        @retry(elasticsearch.exceptions.ConnectionError, tries=TIMES_TO_TRY)
        def next_scroll(scroll_id):
            return self.es_conn.scroll(scroll=self.scroll_time, scroll_id=scroll_id)

        search_args = dict(
            index=','.join(self.opts.index_prefixes),
            sort=','.join(self.opts.sort),
            scroll=self.scroll_time,
            size=self.opts.scroll_size,
            terminate_after=self.opts.max_results
        )

        if self.opts.query.startswith('@'):
            query_file = self.opts.query[1:]
            if os.path.exists(query_file):
                with codecs.open(query_file, mode='r', encoding='utf-8') as f:
                    self.opts.query = f.read()
            else:
                print('No such file: {}.'.format(query_file))
                exit(1)
        
        try:
            query = json.loads(self.opts.query)
        except ValueError as e:
            print('Invalid JSON syntax in query. {}'.format(e))
            exit(1)
        search_args['body'] = query
        
        if '_all' not in self.opts.fields:
            search_args['_source'] = ','.join(self.opts.fields)

        if self.opts.debug_mode:
            print('Using these indices: {}.'.format(', '.join(self.opts.index_prefixes)))
            print('Query[{0[0]}]: {0[1]}.'.format(
                ('Query DSL', json.dumps(query, ensure_ascii=False).encode('utf8')) if self.opts.raw_query else ('Lucene', query))
            )
            print('Output field(s): {}.'.format(', '.join(self.opts.fields)))
            print('Sorting by: {}.'.format(', '.join(self.opts.sort)))

        res = self.es_conn.search(**search_args)
        self.num_results = res['hits']['total']['value']

        print('Found {} results.'.format(self.num_results))
        if self.opts.debug_mode:
            print(json.dumps(res, ensure_ascii=False).encode('utf8'))

        if self.num_results > 0:
            codecs.open(self.file, mode='w', encoding='utf-8').close()
            
            if res['hits']['hits']:
                if '_source' in res['hits']['hits'][0] and len(res['hits']['hits'][0]['_source']) > 0:
                    def get_header(source, ancestors=[], header_delimeter='.'):
                        def is_list(arg):
                            return type(arg) is list
                        
                        def is_dict(arg):
                            return type(arg) is dict
                        
                        if is_dict(source):
                            for key in source.keys():
                                get_header(source[key], ancestors + [key])
                        elif is_list(source):
                            [get_header(item, ancestors + [str(index)]) for index, item in enumerate(source)]
                        else:
                            header = header_delimeter.join(ancestors)
                            self.fields.append(header)
            
                    get_header(res['hits']['hits'][0]['_source'])
                    if self.opts.header_on:
                        with codecs.open(self.file, mode='a', encoding='utf-8') as tmp_file:
                            tmp_file.write("\t\t".join(self.fields).rstrip("\n") + "\n")
                            
            hit_list = []
            total_lines = 0

            widgets = ['Run query ',
                       progressbar.Bar(left='[', marker='#', right=']'),
                       progressbar.FormatLabel(' [%(value)i/%(max)i] ['),
                       progressbar.Percentage(),
                       progressbar.FormatLabel('] [%(elapsed)s] ['),
                       progressbar.ETA(), '] [',
                       progressbar.FileTransferSpeed(unit='docs'), ']'
                       ]
            bar = progressbar.ProgressBar(widgets=widgets, maxval=self.num_results).start()

            while total_lines != self.num_results:
                if res['_scroll_id'] not in self.scroll_ids:
                    self.scroll_ids.append(res['_scroll_id'])

                if not res['hits']['hits']:
                    print('Scroll[{}] expired(multiple reads?). Saving loaded data.'.format(res['_scroll_id']))
                    break
                for hit in res['hits']['hits']:
                    total_lines += 1
                    bar.update(total_lines)
                    hit_list.append(hit)
                    if len(hit_list) == FLUSH_BUFFER:
                        self.flush_to_file(hit_list)
                        hit_list = []
                    if self.opts.max_results:
                        if total_lines == self.opts.max_results:
                            self.flush_to_file(hit_list)
                            print('Hit max result limit: {} records'.format(self.opts.max_results))
                            return
                res = next_scroll(res['_scroll_id'])
            self.flush_to_file(hit_list)
            bar.finish()

    def flush_to_file(self, hit_list):
        def to_keyvalue_pairs(source, ancestors=[], header_delimeter='.'):
            def is_list(arg):
                return type(arg) is list

            def is_dict(arg):
                return type(arg) is dict

            if is_dict(source):
                for key in source.keys():
                    to_keyvalue_pairs(source[key], ancestors + [key])

            elif is_list(source):
                [to_keyvalue_pairs(item, ancestors + [str(index)]) for index, item in enumerate(source)]
            else:
                header = header_delimeter.join(ancestors)
                out[header] = source

        with codecs.open(self.file, mode='a', encoding='utf-8') as tmp_file:
            for hit in hit_list:
                if '_source' in hit and len(hit['_source']) > 0:
                    out = {}
                    to_keyvalue_pairs(hit['_source'])
                    tmp_file.write("\t\t".join([str(out[field]) for field in self.fields]).rstrip("\n") + "\n")
        tmp_file.close()

    def clean_scroll_ids(self):
        try:
            for sid in self.scroll_ids:
                self.es_conn.clear_scroll(scroll_id=sid)
        except Exception as e:
            print(str(e))
            
    def exit_by_signal(self, signum, frame):
        self.clean_scroll_ids()
        
    def compress_file(self):
        gz_file = self.file + '.gz'
        with open(self.file, "rb") as f_in:
            with gzip.open(gz_file, "wb") as f_out:
                f_out.write(f_in.read())
        os.remove(self.file)


esExporter_cli.py
#!/usr/bin/env python
"""
title:           A CLI tool for exporting data from Elasticsearch.
description:     Baseed on https://github.com/taraslayshchuk/es2csv developed. Command line utility, for querying Elasticsearch(7.x) in Lucene query syntax or Query DSL syntax and exporting result as documents into a file with custom fields.
usage:           esExporter -q '*' -i *idx* -o ~/file.txt -m 100
                 esExporter -q '{"query": {"match_all": {}}}' -i *idx* -o ~/file.txt
                 esExporter -q '{"query": {"match_all": {}}}' -i *idx* -m 10 --header-on -o ~/file.txt
                 esExporter -q @'~/long_query_file.json' -i *idx* -o ~/file.txt
                 esExporter -q '*' -i *idx* -f @timestamp log kubernetes.host -o ~/file.txt -m 100
                 esExporter -q '{"query": {"match_all": {}}, "filter":{"term":{"tags":"dev"}}}' -u http://elasticsearch-master:9200 -o ~/file.txt
"""
import sys
import argparse
import esExporter

__version__ = '7.5.1'


def main():
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument('-q', '--query', dest='query', type=str, required=True, help='Query string in Lucene syntax.')
    p.add_argument('-u', '--url', dest='url', default='http://localhost:9200', type=str, help='Elasticsearch host URL. Default is %(default)s.')
    p.add_argument('-a', '--auth', dest='auth', type=str, required=False, help='Elasticsearch basic authentication in the form of username:password.')
    p.add_argument('-i', '--index-prefixes', dest='index_prefixes', default=['logstash-*'], type=str, nargs='+', metavar='INDEX', help='Index name prefix(es). Default is %(default)s.')
    p.add_argument('-o', '--output-file', dest='output_file', type=str, required=True, metavar='FILE', help='CSV file location.')
    p.add_argument('-f', '--fields', dest='fields', default=['_all'], type=str, nargs='+', help='List of selected fields in output. Default is %(default)s.')
    p.add_argument('-S', '--sort', dest='sort', default=[], type=str, nargs='+', metavar='FIELDS', help='List of <field>:<direction> pairs to sort on. Default is %(default)s.')
    p.add_argument('-d', '--delimiter', dest='delimiter', default=',', type=str, help='Delimiter to use in CSV file. Default is "%(default)s".')
    p.add_argument('-m', '--max', dest='max_results', default=0, type=int, metavar='INTEGER', help='Maximum number of results to return. Default is %(default)s.')
    p.add_argument('-s', '--scroll-size', dest='scroll_size', default=1000, type=int, metavar='INTEGER', help='Scroll size for each batch of results. Default is %(default)s.')
    p.add_argument('--verify-certs', dest='verify_certs', action='store_true', help='Verify SSL certificates. Default is %(default)s.')
    p.add_argument('--ca-certs', dest='ca_certs', default=None, type=str, help='Location of CA bundle.')
    p.add_argument('--client-cert', dest='client_cert', default=None, type=str, help='Location of Client Auth cert.')
    p.add_argument('--client-key', dest='client_key', default=None, type=str, help='Location of Client Cert Key.')
    p.add_argument('-v', '--version', action='version', version='%(prog)s ' + __version__, help='Show version and exit.')
    p.add_argument('--header-on', dest='header_on', action='store_true', help='Export file with header.')
    p.add_argument('--debug', dest='debug_mode', action='store_true', help='Debug mode on.')

    if len(sys.argv) == 1:
        p.print_help()
        exit()

    opts = p.parse_args()
    es = esExporter.EsExporter(opts)
    es.create_connection()
    es.check_indexes()
    es.search_query()
    es.clean_scroll_ids()
    es.compress_file()

if __name__ == '__main__':
    main()
